Needs testing: Rewrite crafting skill checks to use normal_roll.

I-am-Erk opened the pull request:
#### Summary
SUMMARY: Infrastructure "Rewrite crafting skill checks to use normal_roll"

#### Purpose of change
I have revived this PR after a few years inactive, because we still need it and it turns out it wasn't that stale.

~~1. Proficiency failure penalties are currently not being applied correctly. This is indirectly because the way skill checks are calculated is needlessly complicated.~~ no longer relevant.
1. Per #44738, it is a good idea to start moving major skill checks into normal rolls. Then when we create a standardized roll_check routine, it will be much easier to write and to migrate. The current crafting skill check method is impenetrably complex and the complexity offers no real benefit.
2. Per #66064, there are two crafting check issues that need resolution: first, crafting checks from skill 0 always fail, and second, crafting checks do not take into account marginal skill gain which interacts poorly with rust mechanics and will cause problems when skill gain is slowed.

#### Describe the solution
To solve both of these, I have migrated the crafting skill check to use normal_roll and make it mathematically predictable. This brings in a few major changes:

1. NPCs no longer have to be more skilled than you to increase your chance of success. Instead, each NPC able to help that also has all the prerequisites to be able to do the craft on their own gets a skill check just like you. Then we add those skill checks together using the formula: 
```
final skill value = sqrt( your skill value^2 + sum (all contributing NPC skill values^2) )
```
This root-sum-of-squares is the same system used in summing morale: it creates diminishing returns the more contributors you have, and makes more skilled sources of help substantially outweigh smaller ones. Using this formula, for example, if you and a helper NPC both have an effective skill level of 7 once all effects are applied, your combined skill level is around 10. If you add another NPC of effective skill level 7, you get a total combined skill of 12.

In the end this makes NPC helpers much more useful (previously they just added a small flat bonus, and only if they were more skilled than you).

3. Changes the role of secondary and primary skills in your check, making secondary skills potentially more important. Previously the weight of primary and secondary skill requirements was flat: 3/4 of your skill check was from your primary skill and 1/4 was from the average of the secondary skills. Now it is a weighted average of: `2*primary skill*required level of primary skill, + sum of secondary skills * required level of secondary skills`. That means if a recipe asks for electronics 2 and cooking 1 (what the hell recipe is this??), your overall level in electronics will be twice as important to your final skill value. If that same recipe uses fabrication 2 as its primary skill, then fabrication will weigh in at 2x as important as electronics and 4x as important as cooking.
4. Moves the method of calculating your effective skill level to a new method to make it easier to determine your and your NPCs' effective skills with all modifiers included.
5. **(not done)** Includes your marginal skill gain in the calculation, so having skill level 1(50%) is equal to having skill level 1.5, and progress in a skill matters.


#### Describe alternatives you've considered
I would like some more features in a follow up PR.
- I want to jsonize what role attributes have in skills. Currently only intelligence matters. I'd suggest the simplest way would be to define defaults in the skills in skills.json themselves, and add some optional fields for recipes that could override these defaults at some later date, similar to how we do it for proficiencies where there's a default and an override. So for example, electronics recipes could be intelligence based by default, but often use perception instead.
    - If we decide to move away from using attributes at all in the next stable, this should instead use the surrogates we develop for attribute roles. 
- I want to jsonize mutations that affect specific skills. Since I first did this PR the mutations themselves have been jsonized, but I would like the skill penalty to be in JSON, either as part of the mutation or part of the skill.
- I would very much like to display chance to craft in the crafting UI, but I doubt I have the skill to do that.


#### Testing
Compiles, loads, and allows crafting. Crafting skill rolls appear to be happening correctly but do have too wide of a range in my opinion. Failures seem to be a bit difficult still, but I would like to test more and I would like some help with that.

![image](https://user-images.githubusercontent.com/45136638/184281009-5d000a17-3931-436d-a8c9-54fb95e9029b.png)

As you can see this has my second roll on an effective skill of 3 as a 10. That might just be an outlier but it's pretty sus, I think the standard deviation widening at lower levels may be a mistake, or perhaps that should go along with also dropping the roll a bit.

#### Additional context
It's kind of insane that this PR loaded and compiled with minimal edits after 2 years.

I-am-Erk committed:
Pretty much totally rewrite crafting skill checks. Not tested.

actual-nh commented on src/character.h:
```diff
@@ -2500,6 +2500,7 @@ class Character : public Creature, public visitable<Character>
          * @return a value >= 0.0 with >= 1.0 representing unequivocal success
          */
         double crafting_success_roll( const recipe &making ) const;
+        float get_recipe_weighted_skill_average( const recipe &making )
```
```suggestion
        float get_recipe_weighted_skill_average( const recipe &making );
```

I-am-Erk committed:
Update src/character.h

Co-authored-by: actual-nh <74678550+actual-nh@users.noreply.github.com>

I-am-Erk committed:
fix exponent

I-am-Erk committed:
fix use of focus pool

I-am-Erk committed:
a couple more small fixes

I-am-Erk committed:
apply @anothersimulacrum patch

I-am-Erk committed:
Update crafting.cpp

I-am-Erk committed:
Update crafting.cpp

Make intelligence neutral at an average of 8.

I-am-Erk committed:
Update crafting.cpp

make final_difficulty tallied the same as skill total
fix a dumb gaffe from me

I-am-Erk committed:
fix variable name

I-am-Erk committed:
account for possibly negative player skill

I-am-Erk committed:
dept of redundancy dept

I-am-Erk committed:
adjust standard deviation based on how skilled you are

actual-nh commented on src/crafting.cpp:
```diff
@@ -1055,6 +1052,20 @@ double Character::crafting_success_roll( const recipe &making ) const
         ( 2.0f * making.difficulty + 1.0f * secondary_level_count );
 
+    // in the future we might want to make the standard deviation vary depending on some feature of the recipe.
+    float crafting_stddev = 2.0f
```
```suggestion
    float crafting_stddev = 2.0f;
```

anothersimulacrum commented on src/crafting.cpp:
```diff
@@ -1055,6 +1052,20 @@ double Character::crafting_success_roll( const recipe &making ) const
+        // This means that luck plays less of a roll the more overqualified you are.
+        crafting_stddev -= std::min( ( weighted_skill_average - final_difficulty ) / 4, 1.0f );
+    }
+    float craft_roll = std::max( normal_roll( weighted_skill_average, crafting_stddev ), 0.0f );
```
```suggestion
    float craft_roll = std::max( normal_roll( weighted_skill_average, crafting_stddev ), 0.0 );
```

actual-nh commented on src/crafting.cpp:
```diff
@@ -940,86 +940,133 @@ void Character::craft_proficiency_gain( const item &craft, const time_duration &
+    }
+    if( final_difficulty < weighted_skill_average ){
+        // decrease the standard deviation by 0.25 for every point above the difficulty your skill level is, to a cap of 1.
+        // This means that luck plays less of a roll the more overqualified you are.
```
```suggestion
        // This means that luck plays less of a role the more overqualified you are.
```
Unless a pun!

I-am-Erk committed:
Update src/crafting.cpp

Co-authored-by: actual-nh <74678550+actual-nh@users.noreply.github.com>

I-am-Erk committed:
Update src/crafting.cpp

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>

I-am-Erk committed:
Update src/crafting.cpp

Co-authored-by: actual-nh <74678550+actual-nh@users.noreply.github.com>

Kelenius commented on src/crafting.cpp:
```diff
@@ -940,86 +940,133 @@ void Character::craft_proficiency_gain( const item &craft, const time_duration &
-    // Sides on dice is 16 plus your current intelligence
-    ///\EFFECT_INT increases crafting success chance
-    const int skill_sides = 16 + int_cur;
+    // Focus has a role. If it is below 100, it is a penalty to success.  Over 100, it is a bonus.
```
Why focus? It's finicky and goes down for reasons absolutely not related to crafting. With this change, reading a book on electronics before crafting something would increase the change of failure, but staring at a wall for 30 minutes will decrease it.

I-am-Erk replied:
You have a point. For the moment, until focus can better represent 'mental stamina', morale would probably do the job better.

actual-nh replied:
We seem to be thinking alike on this - see #48758.

pehamm commented on src/crafting.cpp:
```diff
@@ -940,86 +940,133 @@ void Character::craft_proficiency_gain( const item &craft, const time_duration &
+    }
+    float craft_roll = std::max( normal_roll( weighted_skill_average, crafting_stddev ), 0.0 );
+    
+    return std::max( craft_roll / final_difficulty, 0.0f );
```
May I suggest we switch from a ratio of two normals to the difference of two normals? The ratio of two normals is well known to be a mess to work with. This would require some changes as the `max(A/B, 0)` would no longer be appropiate and success would have to be redefined as `success_roll > 0.0` instead of `success_roll > 1.0`, but it would have the advantage that we would have a behaving output distribution. This would enable relatively straightforward computation of probabilities so we could show to people in the crafting menu what their failure probability is.

I-am-Erk replied:
Yes, we should do that.

I-am-Erk replied:
I believe the reason it is presently done like this is to attempt to maintain the same range output results so I don't have to alter the failure case stuff. However it might be easy enough to do that with subtraction instead of division.

actual-nh replied:
Divide the result by the average output of the difficulty dice (or the skill dice if preferable), then add 1?

I-am-Erk replied:
There aren't really difficulty or skill dice anymore. there is a normal roll, centered around your average skill with bonuses/penalties, and a constant difficulty determined by all contributing skills.

It might be possible to just subtract A from B and apply it to current rolls. Presently they're based on a ratio of less than or greater than 1 though so I guess we probably need to add 1 to that final value such that if your success perfectly equals the difficulty, that's considered a pass.

actual-nh replied:
> There aren't really difficulty or skill dice anymore. there is a normal roll, centered around your average skill with bonuses/penalties, and a constant difficulty determined by all contributing skills.

Got it; was confused by how it was previously set up, sorry.

> It might be possible to just subtract A from B and apply it to current rolls. Presently they're based on a ratio of less than or greater than 1 though so I guess we probably need to add 1 to that final value such that if your success perfectly equals the difficulty, that's considered a pass.

How about (for the purpose of reusing existing code that doesn't seem to be problematic, like getting proportions of lost components and progress, and potentially for figuring out any additional advantages of a very successful roll, like things getting done faster than normal):
- Take roll - difficulty:
- - If this is above 0, divide by the difficulty. (Thus, the greater the difficulty, the harder it is to get any additional advantages.)
- - If this is below 0, divide by the _mean_ result of the roll. (Thus, the better your skill, the more limited any disaster due to mischance is.)
- Add 1.
- Take max(this, 0).

I-am-Erk replied:
The code actually already accounts for your likelihood of avoiding disaster at higher levels by reducing the standard deviation, and expanding it at skill deficit. so at high levels you get a progressively less broad spread of results.

I'll have to review the rest of your math there. I think I'd prefer to have a simple calculation and go through and edit how it's used elsewhere than have a complex calculation that can be plugged I to the existing code seemlessly though.

I-am-Erk committed:
Update crafting.cpp

just some comments as I review.

I-am-Erk committed:
Update crafting.cpp

I-am-Erk committed:
Merge branch 'master' into change-skill-success-rolls

I-am-Erk committed:
Update crafting.cpp

pehamm commented:
Fuck, I was staring for too long at the current code and did not realize final_difficulty is no longer random. If final_difficulty is a constant, taking the difference versus the ratio does not matter, the output stays normal. Only if the difficulty were a bell curve (like it is now on the master) does the ratio lead to a "Cauchy"-like fat tailed distribution while the difference stays normal. Sorry for the confusion!

If you want, I can make some simulation runs and supply results (similar to what I did here: https://github.com/CleverRaven/Cataclysm-DDA/issues/48981#issuecomment-847150711), to give a better estimate of what is happening. That's something I generally do on the job whenever I have to work with stochastic processes (which is often).

We have a lot of people complaining about success rates when proficiencies are missing. Part of it is that many misunderstand the term fail multiplier, which you are already looking at. I would have suggested replacing the formula in a way that 2x failure multiplier means "twice the average number of failures until crafting success", but skill penalty works just as well.

I-am-Erk commented:
> Fuck, I was staring for too long at the current code and did not realize final_difficulty is no longer random. If final_difficulty is a constant, taking the difference versus the ratio does not matter, the output stays normal. Only if the difficulty were a bell curve (like it is now on the master) does the ratio lead to a "Cauchy"-like fat tailed distribution while the difference stays normal. Sorry for the confusion!
> 
> If you want, I can make some simulation runs and supply results (similar to what I did here: [#48981 (comment)](https://github.com/CleverRaven/Cataclysm-DDA/issues/48981#issuecomment-847150711)), to give a better estimate of what is happening. That's something I generally do on the job whenever I have to work with stochastic processes (which is often).
> 
> We have a lot of people complaining about success rates when proficiencies are missing. Part of it is that many misunderstand the term fail multiplier, which you are already looking at. I would have suggested replacing the formula in a way that 2x failure multiplier means "twice the average number of failures until crafting success", but skill penalty works just as well.

I was pretty sure you were making that error, but unless I'm sorely mistaken, I think it's still a better calculation to use subtraction rather than division. It's a lot more intuitive.

I'd be deeply grateful for someone to run some simulations. I tend to be intuitively good at this stuff and I've done some rough testing, but nothing very exhaustive, and the widening normal curve at high and low skill levels is something that definitely needs to be examined more closely.

Before taking this out of draft I currently plan to rename "fail multiplier" to "skill penalty", and make it work as such, which should clear up any confusion.

I-am-Erk committed:
Update crafting.cpp

begone, unused variable.

I-am-Erk commented:
I will pick this up again when i have time and energy.

I-am-Erk committed:
Merge branch 'master' into change-skill-success-rolls

I-am-Erk committed:
remove focus from equation for now

can be worked to use morale later, not important at the moment.

I-am-Erk commented on src/crafting.cpp:
```diff
@@ -993,82 +993,137 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    for( const recipe_proficiency &recip : making.proficiencies ) {
+        if( !recip.required && !has_proficiency( recip.id ) ) {
+            total_skill_modifiers -= ( recip.fail_multiplier - 1.0f );
+        }
```
This is outdated now, this would have worked for proficiency fail mults in the 1-5 range, but with them ranging 1-1.5 this is not going to do much of anything. I can either redo all the fail multipliers again or just move them elsewhere, will probably do hte latter.

I-am-Erk commented on src/crafting.cpp:
```diff
@@ -993,82 +993,137 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    // weight added by the required level.
+    const float weighted_skill_average =
+        ( ( 2.0f * making.difficulty * get_skill_level( making.skill_used ) ) + secondary_skill_total ) /
+        ( 2.0f * making.difficulty + secondary_difficulty );
```
If we want to make the fail multipliers on proficiencies into a real multiplier, here is where we'd add them, by dividing your `weighted_skill_average` by `recip.fail_multiplier`.

I-am-Erk committed:
comments

I-am-Erk committed:
astyle

I-am-Erk committed:
add debug messages

I-am-Erk commented on src/crafting.cpp:
```diff
@@ -993,82 +993,148 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
 {
     if( has_trait( trait_DEBUG_CNF ) ) {
         return 1.0;
     }
```
pretty sure this part should still be in crafting_success_roll() not here

I-am-Erk commented on src/crafting.cpp:
```diff
@@ -993,82 +993,148 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    float summed_skills = player_weighted_skill_average + npc_helpers_weighted_average;
+    if( summed_skills < 0 ) {
+        negative = true;
+        summed_skills *= -1;
```
```suggestion
```
This is a bad idea, this will make is take the sqrt of a negative and we are not trying to make a fractal.

I-am-Erk committed:
Update src/crafting.cpp

I-am-Erk commented on src/crafting.cpp:
```diff
@@ -993,82 +993,147 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    negative = false;
+    float summed_skills = player_weighted_skill_average + npc_helpers_weighted_average;
+    if( summed_skills < 0 ) {
+        negative = true;
```
```suggestion
        negative = true;
        // Store that it was negative to reapply later, then make it positive because we need a sqrt.
        summed_skills *= -1;
```
I was right the first time, math is hard.

I-am-Erk committed:
Update src/crafting.cpp

I-am-Erk committed:
Update crafting.cpp

I-am-Erk commented:
Shockingly and against all odds this is now at a place where people could help me playtest. It is not yet ready to merge.

I-am-Erk committed:
astyle

mqrause commented on src/crafting.cpp:
```diff
@@ -993,82 +993,150 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    // weight added by the required level.
+    const float weighted_skill_average =
+        ( ( 2.0f * making.difficulty * get_skill_level( making.skill_used ) ) + secondary_skill_total ) /
+        ( 2.0f * making.difficulty + secondary_difficulty );
```
This causes division by 0 for 0 difficulty recipes.

I-am-Erk replied:
thanks, I'll fix that

I-am-Erk commented on src/crafting.cpp:
```diff
@@ -993,82 +993,150 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    // weight added by the required level.
+    const float weighted_skill_average =
+        ( ( 2.0f * making.difficulty * get_skill_level( making.skill_used ) ) + secondary_skill_total ) /
+        ( 2.0f * making.difficulty + secondary_difficulty );
```
```suggestion
        std::max( ( 2.0f * making.difficulty + secondary_difficulty ), 1.0f );
```

I-am-Erk committed:
Update src/crafting.cpp

anothersimulacrum committed:
Add failure chance calculation, adjust catastrophic failure odds (#27)

* Display recipe chance of failure

Split the crafting roll function into just providing the parameters for
the normal curve it will be rolling on, and a function doing the actual
roll.
Also, fix some DBZ bugs, and add debug outputs while we're there.

Then, add a function to calculate the chance we'll pass a normal roll
with given center, stddev, and difficulty, and use that to provide the
recipe success chance.
Add this to the crafting GUI.

* Adjust crafting catastrophic failure chances

Fudge the numbers to make catastrophic (item-destroying) failures much
less likely than setback failures.
Also display the chance of a catastrophic failure in the UI.

Night-Pryanik commented:
@I-am-Erk could you please resolve conflicts?

dseguin committed:
Merge branch 'master' into change-skill-success-rolls

anothersimulacrum commented on src/crafting.cpp:
```diff
@@ -999,82 +999,248 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+// said curve will return a value of difficulty or greater.
+static float normal_roll_chance( float center, float stddev, float difficulty )
+{
+    cata_assert( stddev >= 0.f );
```
```suggestion
    cata_assert( stddev > 0.f );
```
